include/deal.II/physics/notation.h

```diff
@@ -0,0 +1,619 @@
+      unroll_to_matrix (const Tensor<1,dim,Number> &v);
+
+
+      /**
+       * Convert a rank-2 tensor to its compressed matrix equivalent.
+       */
+      template<typename Number>
+      static
+      FullMatrix<Number>
+      unroll_to_matrix (const Tensor<2,dim,Number> &t);
+
+
+      /**
+       * Convert a rank-2 symmetric tensor to its compressed matrix equivalent.
+       */
+      template<typename Number>
+      static
+      FullMatrix<Number>
+      unroll_to_matrix (const SymmetricTensor<2,dim,Number> &st);
+
+      /**
+       * The template arguments @p SubTensor1 and @p SubTensor2 determine how
+       * the unrolling occurs, in particular how the elements of the rank-3
+       * tensor are to be interpreted.
+       *
+       * So, for example, with the following two conversions
+       * @code
+       * Tensor<3,dim> r3_tnsr;      // All elements filled differently
+       * Tensor<3,dim> r3_symm_tnsr; // Some elements filled symmetrically
+       *
+       * const FullMatrix<double> mtrx_1
+       *   = Physics::Notation::Kelvin<dim>::template unroll_to_matrix< Tensor<2,dim>,Tensor<1,dim> >(r3_tnsr);
+       * const FullMatrix<double> mtrx_2
+       *   = Physics::Notation::Kelvin<dim>::template unroll_to_matrix< Tensor<1,dim>,SymmetricTensor<2,dim> >(r3_symm_tnsr);
+       * @endcode
+       * the matrix @p mtrx_1 will have $dim \times dim$ rows and $dim$ columns
+       * (i.e. size Tensor<2,dim>::n_independent_components $\times$ Tensor<1,dim>::n_independent_components),
+       * while those of the matrix @p mtrx_2 will have $dim$ rows and
+       * $(dim \times dim + dim)/2$ columns
+       * (i.e. size Tensor<1,dim>::n_independent_components $\times$ SymmetricTensor<2,dim>::n_independent_components),
+       * as it is assumed that the entries corresponding to the alternation of the
+       * second and third indices are equal.
+       * That is to say that <code>r3_symm_tnsr[i][j][k] == r3_symm_tnsr[i][k][j]</code>.
+       */
+      template<typename SubTensor1 = Tensor<2,dim>,
+               typename SubTensor2 = Tensor<1,dim>,
+               typename Number>
+      static
+      FullMatrix<Number>
+      unroll_to_matrix (const Tensor<3,dim,Number> &t);
+
+
+      /**
+       * Convert a rank-4 tensor to its compressed matrix equivalent.
+       */
+      template<typename Number>
+      static
+      FullMatrix<Number>
+      unroll_to_matrix (const Tensor<4,dim,Number> &t);
+
+
+      /**
+       * Convert a rank-4 symmetric tensor to its compressed matrix equivalent.
+       */
+      template<typename Number>
+      static
+      FullMatrix<Number>
+      unroll_to_matrix (const SymmetricTensor<4,dim,Number> &st);
+
+//@}
+
+      /**
+       * @name Reverse operation: Kelvin notation to tensor notation
+       */
+//@{
+
+      /**
+       * Convert a compressed vector to its equivalent scalar value.
+       */
+      template<typename Number>
+      static void
+      roll_to_tensor (Number &s,
```

Comment:
any reason not to return the result, to be consistent with the above? We shall have move constructors now for Tensors, so it should be ok

Reply:
The compiler can't automatically deduce the result type. Thats what the convenience functions, e.g. [here](https://github.com/dealii/dealii/pull/5250/files/36e9f33fe51d74d95f4917fd432811218cec4549#diff-e9c933ece0fa2c81679a92f51415cfd2R525), are for.